### DISCUSSION TITLE
ArgsKwargs in V2 pydantic.dataclasses

### DISCUSSION THREAD

**Discussion by bestrauc** [pydantic/pydantic#9939] [Question]
[This code in docs](https://docs.pydantic.dev/latest/concepts/dataclasses/#initialization-hooks) has an example of dataclass validation:

```python
@dataclass
class User:
    birth: Birth

    @model_validator(mode='before')
    @classmethod
    def pre_root(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        print(f'First: {values}')
        """
        First: ArgsKwargs((), {'birth': {'year': 1995, 'month': 3, 'day': 2}})
        """
        return values
```

Can someone elaborate why `values` is of type `ArgsKwargs` instead of `Dict[str, Any]` as indicated in the type annotation? In this code, I wouldn't be able to validate things by accessing `values["birth"]`, because the values are in fact not a dictionary.

If the class is defined via a `BaseModel` instead of pydantic dataclass, the behavior is normal and you do get a dictionary.

I'm currently running into this while transitioning from V1 to V2, because previously the V1 `@root_validator` worked as expected and after the conversion to `@model_validator`, `values` unfortunately changed its type to the basically undocumented `ArgsKwargs`.

Am I expected to destructure this into args & kwargs in the body of the validator in case of dataclasses? This is of course a workaround, but the lack of documentation on `ArgsKwargs` left me wondering.

**Comment by cmclaughlin** [comment]
@bestrauc I experienced this too... although there are no answers here, I'm wondering if you figured it out on your own.

**Comment by cmclaughlin** [comment]
This seems to work...

```
values.kwargs
```

**Reply by cmclaughlin** [reply]
i.e. in that example, year/month/day are in `values.kwargs["birth"]`

**Comment by bestrauc** [comment]
Hey @cmclaughlin , I didn't end up investigating further, but I also think that accessing `values.kwargs` is probably the thing I'd have to fall back on.

Still weird that the docs are inconsistent and this `ArgsKwargs` object isn't really documented anywhere. Since a non-dataclass `BaseModel` works correctly, it feels like this is a bug with Pydantic dataclasses.
